src/main/java/org/javarosa/core/model/ItemsetBinding.java

```diff
@@ -56,54 +70,198 @@
+        if (nodesetRef.getInstanceName() != null) { // the itemset is defined in a secondary instance
+            formInstance = formDef.getNonMainInstance(nodesetRef.getInstanceName());
+            if (formInstance == null) {
+                throw new XPathException("Instance " + nodesetRef.getInstanceName() + " not found");
+            }
+        } else {
+            formInstance = formDef.getMainInstance();
         }
-        this.choices = randomize ? shuffle(choices, resolveRandomSeed(model, ec)) : choices;
 
+        List<TreeReference> filteredItemReferences = nodesetExpr.evalNodeset(formDef.getMainInstance(),
+            new EvaluationContext(formDef.getEvaluationContext(), contextRef.contextualize(curQRef)));
+
+        if (filteredItemReferences == null) {
+            throw new XPathException("Could not find references depended on by" + nodesetRef.getInstanceName());
+        }
+
+        Map<String, Boolean> currentAnswersInNewChoices = initializeCurrentAnswerMap(formDef, curQRef);
+
+        List<SelectChoice> choices = new ArrayList<>();
+        for (int i = 0; i < filteredItemReferences.size(); i++) {
+            SelectChoice choice = getChoiceForTreeReference(formDef, formInstance, i, filteredItemReferences.get(i));
+            choices.add(choice);
+            if (currentAnswersInNewChoices != null && currentAnswersInNewChoices.containsKey(choice.getValue())) {
+                currentAnswersInNewChoices.put(choice.getValue(), true);
+            }
+        }
+
+        updateQuestionAnswerInModel(formDef, curQRef, currentAnswersInNewChoices);
+
+        latestFilteredChoiceList = randomize ? shuffle(choices, currentRandomizeSeed) : choices;
+
+        // TODO: write a test that fails if this is removed. It looks like a no-op because it's not accessing the shuffled collection.
         if (randomize) {
             // Match indices to new positions
             for (int i = 0; i < choices.size(); i++)
                 choices.get(i).setIndex(i);
         }
 
         //init localization
-        if (localizer != null) {
-            String curLocale = localizer.getLocale();
+        // TODO: write a test that fails if this is removed
+        if (formDef.getLocalizer() != null) {
+            String curLocale = formDef.getLocalizer().getLocale();
             if (curLocale != null) {
-                localeChanged(curLocale, localizer);
+                localeChanged(curLocale, formDef.getLocalizer());
+            }
+        }
+
+        latestTriggerValues = currentTriggerValues;
+        latestRandomizeSeed = currentRandomizeSeed;
+
+        return latestFilteredChoiceList;
+    }
+
+    /**
+     * Returns a map:
+     *  - keys: the references that are triggers for the nodeset expression
+     *  - values: current values at those references
+     *
+     * Returns null if the nodeset expression has any triggers that are unbounded references because there's no single
+     * value we could track in that case.
```

Comment:
Something’s bothering me about using null to indicate this. But maybe I don’t have an improvement using Java 7. I assume returning an empty map would have a different meaning. A checked exception, maybe?

Reply:
I’d like to better understand how this is tested. Here are the call stacks where null is returned.
```
Called with /data/select[0] from selectFromRepeat_whenRepeatAdded_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatAdded_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatAdded_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatAdded_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatRemoved_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatRemoved_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatRemoved_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatRemoved_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatChanged_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatChanged_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatChanged_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatChanged_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatChanged_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_whenRepeatChanged_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectWithRepeatAsTrigger_recomputesChoiceListAtEveryRequest -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectWithRepeatAsTrigger_recomputesChoiceListAtEveryRequest -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectWithRepeatAsTrigger_recomputesChoiceListAtEveryRequest -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectWithRepeatAsTrigger_recomputesChoiceListAtEveryRequest -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectWithRepeatAsTrigger_recomputesChoiceListAtEveryRequest -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectWithRepeatAsTrigger_recomputesChoiceListAtEveryRequest -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_withPredicate_whenPredicateTriggerChanges_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_withPredicate_whenPredicateTriggerChanges_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_withPredicate_whenPredicateTriggerChanges_updatesChoices -> Scenario.choicesOf -> FormEntryPrompt.getAnswerValue -> FormEntryPrompt.getSelectChoices -> ItemsetBinding.getChoices
Called with /data/select[0] from selectFromRepeat_withPredicate_whenPredicateTriggerChanges_updatesChoices -> Scenario.choicesOf -> ItemsetBinding.getChoices
```

Reply:
We discussed and have an improvement.